[wasm-mt] A version of LowLevelLifoSemaphore that uses callbacks on the browser

lambdageek opened the pull request:
This is Part 2 of https://github.com/dotnet/runtime/issues/84489 - landing support for async JS interop on threadpool threads in multi-threaded WebAssembly.

This PR adds two things:

1. A new unmanaged `LifoSemaphoreAsyncWait` semaphore that can use Emscripten's ability to push C calls from one thread to another in order to implement a callback-based semaphore - when a thread wants to wait, it sets up a success callback and a timeout callback, and then can return to the JS event loop.  When the semaphore is released, Emscripten will trigger the callback to run on the waiting thread.  If the wait times out, the timeout callback will run.
2. A new flavor of the managed `LowLevelLifoSemaphore` that is not allowed to use the normal `Wait()` function, and instead needs to use the callback-based `PrepareAsyncWait()` function.

lambdageek committed:
[wasm-mt][mono] Add new LifoSemaphoreAsyncWait C primitive

Add a new kind of LifoSemaphore that has a callback-based wait
function, instead of a blocking wait using Emscripten's ability to
send work from one webworker to another in C.

This will allow us to wait for a semaphore from the JS event loop in a
web worker.

lambdageek committed:
[wasm-mt][mono] split LowLevelLifoSemaphore into two kinds

A normal LowLevelLifoSemaphore that can do a synchronous wait
and another that can do a callback-based wait from the JS event loop

jkoritzinsky commented on src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:
```diff
@@ -39,6 +39,9 @@ public LowLevelLifoSemaphore(int initialSignalCount, int maximumSignalCount, int
 
         public bool Wait(int timeoutMs, bool spinWait)
         {
+#if TARGET_BROWSER && FEATURE_WASM_THREADS
+            ThrowIfInvalidSemaphoreKind(LifoSemaphoreKind.Normal);
+#endif
```
Are there places where we use `LowLevelLifoSemaphore` that we're keeping on the old "call `Wait`" plan? If not, can we instead `#ifdef`-out the implementation of `Wait` or make it always throw `PlatformNotSupportedException` in this configuration instead of introducing two separate modes that make different sets of method calls valid?

lambdageek replied:
We need both, I think. There are legitimate uses for the synchronous version on the browser.l, too. Just not in the threadpool worker interop scenario.

I could instead do something more OO-y with a common base class that handles the Counters logic. Wasn't sure if it was worthwhile.

jkoritzinsky replied:
I'd prefer that if there isn't a performance issue from doing so. But I know that this is a high perf area, so if it's not possible that's okay.

lambdageek replied:
Made a common base class `LowLevelLifoSemaphoreBase` and two subclasses `LowLevelLifoSemaphore` (same as before) and `LowLevelLifoAsyncWaitSemaphore` (with `PrepareAsyncWait` instead of `Wait`).

lambdageek committed:
Add LowLevelLifoSemaphoreBase

Move the counts to the base class

Move Release to the base class, make ReleaseCore abstract

lambdageek committed:
make a new LowLevelLifoAsyncWaitSemaphore for wasm-mt

lambdageek commented:
There's a weird, but consitent failure on some of the wasm lanes during the build:

```
npm ERR! Cannot read property '@rollup/plugin-typescript' of undefined
  
  npm ERR! A complete log of this run can be found in:
  npm ERR!     /home/cloudtest_azpcontainer/.npm/_logs/2023-04-08T20_46_38_593Z-debug.log
```

I'm not sure if it's something in this PR, or some kind of CI issue.

lambdageek committed:
Revert unintentional package-lock.json changes

jkoritzinsky commented:
I don't have a lot of familiarity with this portion of the Mono code base. I can take a look, but someone from Mono that works on WASM should definitely be the main reviewer for that component.

kouvel reviewed (commented):
Just one question, otherwise LGTM. I only did a quick scan over the native parts, probably would be good to have someone familiar with WASM review as well.

kg commented on src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:
```diff
@@ -0,0 +1,226 @@
+    private void CreateAsyncWait(int maximumSignalCount)
+#pragma warning restore IDE0060
+    {
+        lifo_semaphore = InitInternal(2);
```
named constants for kind please

lambdageek replied:
Actually I think I'm going to make two separate icalls.   `InitInternal(int kind)` is a holdover from before I had two separate subclasses of semaphore.  There's no good reason for it.

kg commented on src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:
```diff
@@ -0,0 +1,226 @@
+    public void PrepareAsyncWait(int timeoutMs, Action<LowLevelLifoAsyncWaitSemaphore, object?> onSuccess, Action<LowLevelLifoAsyncWaitSemaphore, object?> onTimeout, object? state)
+    {
+        //FIXME(ak): the async wait never spins. Shoudl we spin a little?
+        Debug.Assert(timeoutMs >= -1);
```
should we at least check whether the wait can complete instantly before entering the loop? or is that not a scenario that's possible for the lifo semaphore unlike other synchronization primitives

kg commented on src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:
```diff
@@ -0,0 +1,226 @@
+
+        Debug.Fail("unreachable");
+
+#if false
```
why is this here?

lambdageek replied:
I was hoping someone would tell me that it's a good idea to keep the spin-waiting code. I'll delete it.

kouvel replied:
I guess whether to spin-wait or not would depend on perf and other considerations. For instance, is it better to yield back to the JS event loop immediately, or is it better and is it ok to spin-wait a bit for more work and only yield back after the spin-wait fails (potentially the spin-wait may succeed successively and not yield back to the JS event loop for a long time).

lambdageek replied:
> potentially the spin-wait may succeed successively and not yield back to the JS event loop for a long time

I think this is a good reason to hold off on adding spin-waiting here until we see how real code behaves. Potentially starving JS wouldn't be great

kg replied:
I think you can leave this code here for now but not enable it, because the risk of starving the main thread is big yeah

kg commented on src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:
```diff
@@ -0,0 +1,226 @@
+        // if we get here, we need to keep waiting because the SignalCount above was 0 after we did
+        // the CompareExchange - someone took the signal before us.
+        // FIXME(ak): why is the timeoutMs the same as before? wouldn't we starve? why does LowLevelLifoSemaphore.WaitForSignal not decrement timeoutMs?
+        self.PrepareAsyncWaitCore (we.TimeoutMs, we);
```
yeah the timeoutms should be reduced, I think.

lambdageek replied:
@kouvel are you familiar with this part of the (normal) LowLevelLifoSemaphore code: 

https://github.com/dotnet/runtime/blob/44ea019bbbaab1a0bb2877e6510036456a6495eb/src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs#L202-L204

The second time through the outer `while (true)`, if we lost the `InterlockedCompareExchange`, shouldn't `timeoutMs` be decreased by how long we have waited already?

Is it not done that way because measuring the elapsed waiting time is actually too expensive in the common case? (That could certainly be true on wasm, @kg since all the time stuff boils down to an external JS call to `performance.now`)

kouvel replied:
You're right, the timeout should be reduced based on how long the wait took and looks like it's missing in the current implementation. I figure the overhead of timing would be much less than the overhead of WaitCore, though not sure about wasm.

kg replied:
performance.now is a little expensive, but it's still worth doing this right i think.

lambdageek replied:
ok, I'll update the normal LowLevelLifoSemaphore and the async one to decrement `TimeoutMs`

kg commented on src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:
```diff
@@ -0,0 +1,226 @@
+    {
+        GCHandle gchandle = GCHandle.FromIntPtr(userData);
+        WaitEntry internalWaitEntry = (WaitEntry)gchandle.Target!;
+        gchandle.Free();
```
According to the sample in https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.gchandle.fromintptr?view=net-8.0 I'm not sure it's correct to free the handle here. I think it might be a double-free, unless the caller isn't freeing their handle?

lambdageek replied:
Yea this is different from the example in the docs.  In this case the ownership of the GC handle passes from `PrepareAsyncWaitCore` - which creates the GCHandle, to the underlying native wait structure, to the `SuccessCallback` or `TimeoutCallback`. At that point, the callback is the sole owner of the GC Handle (the underlying native wait entry was already freed).  So the callback is responsible for freeing the GCHandle.

In the example in the docs, the call to EnumWindows never owns the GCHandle - it is always owned by the calling function, which is responsible for freeing it.

kg commented on src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.Unix.Mono.cs:
```diff
@@ -5,19 +5,19 @@
+#pragma warning restore IDE0060
         {
-            lifo_semaphore = InitInternal();
+            lifo_semaphore = InitInternal(1);
```
named constant

kg commented on src/mono/mono/utils/lifo-semaphore.c:
```diff
@@ -16,7 +24,7 @@ void
 mono_lifo_semaphore_delete (LifoSemaphore *semaphore)
 {
 	g_assert (semaphore->head == NULL);
```
does this need to occur inside the lock?

kg commented on src/mono/mono/utils/lifo-semaphore.c:
```diff
@@ -80,10 +88,238 @@ mono_lifo_semaphore_release (LifoSemaphore *semaphore, uint32_t count)
+			g_assert (wait_entry->refcount == 2); 
+			--count;
+			/* if we're on the same thread, don't run the callback while holding the lock */
+			emscripten_dispatch_to_thread_async (target, EM_FUNC_SIG_VI, lifo_js_wait_entry_on_success, NULL, wait_entry);
```
is this sufficient to ensure the lock is released before it runs? i assume it posts it to the message queue of that thread?

lambdageek replied:
yea the `_async` variant is like `queueMicrotask`: it will not run the callback immediately, so we will get a chance to unlock.

kg commented:
I don't fully understand everything here, but other than my comments it looks OK

lambdageek committed:
fix possible null dereference

lambdageek committed:
use a separate icall for async wait InitInternal

instead of magic constants that are otherwise not needed in managed

lambdageek committed:
remove dead code; fixup comments

lambdageek committed:
LowLevelLifoSemaphore: decrement timeoutMs if we lost InterlockedCompareExchange

When a thread wakes after waiting for a semaphore to be released, if
it raced with another thread that is also trying to update the
semaphore counts and loses, it has to go back to waiting again.

In that case, decrement the remaining timeout by the elapsed wait time
so that the next wait is shorter.

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:
```diff
@@ -136,13 +136,15 @@ private bool WaitForSignal(int timeoutMs)
 
             while (true)
             {
+                int startTicks = Environment.TickCount;
```
Could avoid timing if the timeout is -1

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:
```diff
@@ -136,13 +136,15 @@ private bool WaitForSignal(int timeoutMs)
                     _separated._counts.InterlockedDecrementWaiterCount();
                     return false;
                 }
+                int elapsedTicks = Environment.TickCount;
```
This seems more like `endTicks` since it's an absolute-time tick value

kouvel commented on src/libraries/System.Private.CoreLib/src/System/Threading/LowLevelLifoSemaphore.cs:
```diff
@@ -173,6 +175,11 @@ private bool WaitForSignal(int timeoutMs)
 
                     counts = countsBeforeUpdate;
+                    if (timeoutMs != -1) {
+                        int waitMs = elapsedTicks - startTicks;
+                        if (waitMs <= timeoutMs)
+                            timeoutMs -= waitMs;
```
Maybe something like this instead would avoid a remote chance of `waitMs` being negative:

```c#
uint waitMs = (uint)(endTicks - startTicks);
if (waitMs < (uint)timeoutMs)
    timeoutMs -= (int)waitMs;
else
    timeoutMs = 0;
```

Also could update `if (!WaitCore(timeoutMs))` at the top of the loop to `if (timeoutMs == 0 || !WaitCore(timeoutMs))` to avoid waiting again with a zero timeout. Maybe similarly for wasm.

kouvel replied:
Could just check for negative `waitMs` too, may be simpler:

```c#
int waitMs = endTicks - startTicks;
if (waitMs >= 0 && waitMs < timeoutMs)
    timeoutMs -= waitMs;
else
    timeoutMs = 0;
```

lambdageek committed:
better timeout decrement code

kouvel commented on src/mono/System.Private.CoreLib/src/System/Threading/LowLevelLifoAsyncWaitSemaphore.Browser.Threads.Mono.cs:
```diff
@@ -112,10 +112,14 @@ private void PrepareAsyncWaitForSignal(int timeoutMs, Action<LowLevelLifoAsyncWa
 
         _onWait();
 
+        if (timeoutMs == 0) {
```
It looks like `InternalAsyncWaitSuccess` calls `PrepareAsyncWaitCore` and not this method, so maybe the timeout check should be in `PrepareAsyncWaitCore`. There is an assert in this method above that the timeout here shouldn't be 0.

lambdageek committed:
move timeoutMs == 0 check to PrepareAsyncWaitCore

make PrepareAsyncWaitCore static and remove a redundant argument

kouvel reviewed (commented):
LGTM, thanks!

lambdageek commented:
@kg @kouvel could I get a sign-off
